lib/documents.js

```diff
@@ -3338,6 +3338,280 @@ function copyConstraints(suggestConstraints, searchConstraints) {
+        }
+  }
+
+  if (transformBatchArray.length !== 0) {
+    console.log(transformBatchArray.length)
+    const batchdef = {documents: transformBatchArray};
+    if(jobState.jobOptions.transform){
+      batchdef.transform = jobState.jobOptions.transform;
+    }
+    transformDocs(jobState, [batchdef],transformBatchArray, writerId);
+  }
+}
+
+function transformDocs(jobState, batchdef,transformBatchArray, writerId){
+
+  jobState.operation.documents = batchdef;
+
+  requester.startRequest(jobState.operation)
+      .result((output) => {
+        jobState.docsTransformedSuccessfully+= transformBatchArray.length;
+        if (jobState.jobOptions.onBatchSuccess) {
+          const progressSoFar = {
+            docsTransformedSuccessfully: jobState.docsTransformedSuccessfully,
+            docsFailedToBeTransformed: jobState.docsFailedToBeTransformed,
+            timeElapsed: (Date.now()-jobState.startTime)
+          };
+          const documents = transformBatchArray;
+          jobState.jobOptions.onBatchSuccess(progressSoFar, documents);
+        }
+        if(jobState.lastBatch){
+          jobState.timeElapsed = Date.now() - jobState.startTime;
+          finishTransformAll(jobState);
+        }
+        else {
+          onTransformAllDocs(jobState, writerId);
+        }
+      })
+      .catch((err) => {
+        if(jobState.error){
+          finishTransformAll(jobState);
+          return;
+        }
+        let errorDisposition = err;
+        if (jobState.jobOptions.onBatchError) {
+          const progressSoFar = {
+            docsTransformedSuccessfully: jobState.docsTransformedSuccessfully,
+            docsFailedToBeTransformed: transformBatchArray.length,
+            timeElapsed: (Date.now()-jobState.startTime)
+          };
+          try {
+            errorDisposition = jobState.jobOptions.onBatchError(progressSoFar, transformBatchArray, err);
+          } catch(err){
+            errorDisposition = err;
+          }
+        }
+
+        if(errorDisposition instanceof Error){
+          jobState.docsFailedToBeTransformed += transformBatchArray.length;
+          jobState.error = errorDisposition;
+          jobState.stream.emit('error', errorDisposition);
+          finishTransformAll(jobState);
+          return;
+        } else if(!Array.isArray(errorDisposition)) {
+          onTransformAllDocs(jobState, writerId);
+          return;
+        }
+
+        switch(jobState.retryCount){
+          case 2:
+          case 1: {
+            jobState.retryCount--;
+            batchdef.documents = errorDisposition;
+            transformDocs(jobState, [batchdef],batchdef.documents, writerId);
+            break;
+          }
+          default: {
+            jobState.docsFailedToBeTransformed += transformBatchArray.length;
+            jobState.error = err;
+            jobState.stream.emit('error', err);
+            finishTransformAll(jobState);
+          }
+        }
+      });
```

Comment:
Could any of the retry logic be refactored into a common function called from multiple error handlers and thus make the implementation easier to maintain?